File: apps/ios/TieComs/Core/Reactions.swift
import Foundation

/// Reglas de las reacciones (mismas que packages/contracts: QUICK_REACTIONS, normalizeEmoji, MAX_REACTIONS_PER_MESSAGE).
enum Reactions {
    /// Barra rápida (mismo orden en web, iOS y Android).
    static let quick = ["👍", "❤️", "😂", "👀", "✅", "🙏"]
    /// 👀 «Lo reviso» y ✅ «Hecho» tienen acción si la empresa la tiene activa.
    static let look = "👀"
    static let done = "✅"
    /// Máximo de emojis distintos por mensaje (el API responde 409 pasado el límite).
    static let maxPerMessage = 20

    /// Emojis comunes del selector completo (el teclado del sistema trae el resto).
    static let common: [String] = [
        "👍", "👎", "❤️", "😂", "🤣", "😊", "😍", "🥰", "😘", "😉", "😎", "🤩", "🥳", "😅", "😆", "🙂",
        "🤔", "🤨", "😐", "🙄", "😏", "😬", "😮", "😯", "😲", "😳", "🥺", "😢", "😭", "😤", "😡", "🤯",
        "😴", "🤒", "🤗", "🤭", "🫡", "🫠", "🙃", "😇", "👀", "✅", "☑️", "✔️", "❌", "⚠️", "❗", "❓",
        "🙏", "👏", "🙌", "💪", "🤝", "👋", "✌️", "🤞", "👌", "🤙", "👉", "👆", "💯", "🔥", "✨", "⭐",
        "🎉", "🎊", "🚀", "💡", "📌", "📎", "📝", "📅", "⏰", "⏳", "💰", "📈", "📉", "🏆", "🎯", "☕",
        "🍕", "🍺", "🥂", "💚", "💙", "💜", "🧡", "💛", "🖤", "🤍", "💔", "🇨🇴", "🇲🇽", "🇵🇪", "🇪🇸", "🇺🇸",
    ]

    /// ¿Es parte de Extended_Pictographic? (Swift no expone la propiedad: isEmoji sin dígitos, #, *,
    /// indicadores regionales ni modificadores de tono).
    static func isPictographic(_ s: Unicode.Scalar) -> Bool {
        let p = s.properties
        guard p.isEmoji else { return false }
        if s.isASCII { return false }
        if (0x1F1E6...0x1F1FF).contains(s.value) { return false }
        if p.isEmojiModifier { return false }
        return true
    }

    private static func isRegional(_ s: Unicode.Scalar) -> Bool { (0x1F1E6...0x1F1FF).contains(s.value) }

    /// Forma canónica (normalizeEmoji del contrato): sin selectores de variación sobrantes y con U+FE0F donde
    /// hace falta (❤ → ❤️, 👍️ → 👍). nil si no es exactamente un emoji.
    static func normalize(_ input: String) -> String? {
        let raw = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty, raw.utf16.count <= 32 else { return nil }
        let vs16 = Unicode.Scalar(0xFE0F)!, zwj = Unicode.Scalar(0x200D)!, keycap = Unicode.Scalar(0x20E3)!
        var parts: [[Unicode.Scalar]] = [[]]
        for s in raw.unicodeScalars where s != vs16 {
            if s == zwj { parts.append([]) } else { parts[parts.count - 1].append(s) }
        }
        let fixed: [[Unicode.Scalar]] = parts.map { part in
            guard let first = part.first else { return part }
            // Tecla: 1⃣ → 1️⃣
            if part.count == 2, "0123456789#*".unicodeScalars.contains(first), part[1] == keycap { return [first, vs16, keycap] }
            let rest = Array(part.dropFirst())
            let skin = rest.first.map { (0x1F3FB...0x1F3FF).contains($0.value) } ?? false
            let needs = isPictographic(first) && !first.properties.isEmojiPresentation && !skin
            return needs ? [first, vs16] + rest : part
        }
        var out = String.UnicodeScalarView()
        for (i, p) in fixed.enumerated() {
            if i > 0 { out.append(zwj) }
            out.append(contentsOf: p)
        }
        let s = String(out)
        guard s.count == 1 else { return nil }
        guard s.unicodeScalars.contains(where: { isPictographic($0) || isRegional($0) || $0 == keycap }) else { return nil }
        return s
    }

    /// Solo emojis (de 1 a 3): el mensaje se muestra grande (isJumbo de la web).
    static func isJumbo(_ body: String) -> Bool {
        let s = body.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !s.isEmpty, s.utf16.count <= 40 else { return false }
        let chars = s.filter { !$0.isWhitespace }
        guard (1...3).contains(chars.count) else { return false }
        let keycap = Unicode.Scalar(0x20E3)!
        return chars.allSatisfy { ch in ch.unicodeScalars.contains { isPictographic($0) || isRegional($0) || $0 == keycap } }
    }

    /// Lista tras poner (`on`) o quitar mi reacción (optimista, como client.react de la web).
    static func toggled(_ list: [ReactionDTO], emoji: String, me: String, on: Bool) -> [ReactionDTO] {
        var next = list.map { r -> ReactionDTO in
            var x = r
            if r.emoji == emoji { x.userIds.removeAll { $0 == me } }
            return x
        }
        if on {
            if let i = next.firstIndex(where: { $0.emoji == emoji }) { next[i].userIds.append(me) }
            else { next.append(ReactionDTO(emoji: emoji, userIds: [me])) }
        }
        return next.filter { $0.count > 0 }
    }

    /// Recordatorio de 👀: en 3 horas, o mañana a las 9:00 si eso cae de noche (lookRemindAt de la web).
    static func lookRemindAt(_ now: Date = Date(), calendar: Calendar = .current) -> Date {
        let at = now.addingTimeInterval(3 * 3600)
        if calendar.component(.hour, from: at) >= 19 || !calendar.isDate(at, inSameDayAs: now) {
            let tomorrow = calendar.date(byAdding: .day, value: 1, to: now) ?? at
            return calendar.date(bySettingHour: 9, minute: 0, second: 0, of: tomorrow) ?? at
        }
        return at
    }

    /// «Laura (Xertify), Beto y Pedro · WhatsApp» (reactorsText de la web).
    static func reactorsText(_ d: BootstrapDTO, _ r: ReactionDTO) -> String {
        var names = r.userIds.map { u -> String in
            if u == d.me.id { return L("common.youShort") }
            guard let p = Naming.person(d, u) else { return L("chat.formerParticipant") }
            if let o = Naming.org(d, p.orgId), o.id != d.me.primaryOrgId { return "\(p.name) (\(o.name))" }
            return p.name
        }
        names += r.external.map { "\($0.name) · \(L("src.\($0.source.rawValue)"))" }
        guard names.count > 1 else { return names.first ?? "" }
        return names.dropLast().joined(separator: ", ") + " \(L("common.and")) " + names.last!
    }

    /// 👀 y ✅ con acción para la gente de mi empresa (activas por defecto).
    static func actionsEnabled(_ d: BootstrapDTO) -> Bool {
        Naming.org(d, d.me.primaryOrgId)?.reactionActions ?? true
    }
}

/// Respuesta de PUT/DELETE /messages/:id/reactions/:emoji.
struct ReactionResult: Decodable {
    var message: MessageDTO?
    var reminder: ReminderDTO?
    var closedReminderIds: [String]
    var openIssueId: String?

    init(from decoder: Decoder) throws {
        let c = try container(decoder)
        message = c.o("message")
        reminder = c.o("reminder")
        closedReminderIds = c.v("closedReminderIds", [])
        openIssueId = c.o("openIssueId")
    }
}

extension AppStore {
    /// Pone o quita mi reacción (optimista; se revierte si el API falla). 👀 manda la hora del recordatorio.
    @discardableResult
    func react(_ m: MessageDTO, emoji raw: String, on: Bool) async throws -> ReactionResult {
        guard let emoji = Reactions.normalize(raw), let d = data else {
            throw ApiRequestError(status: 400, code: "invalid_emoji", message: L("common.error"))
        }
        let current = conversations[m.conversationId]?.messages.first { $0.id == m.id } ?? m
        let prev = current.reactions
        var next = current
        next.reactions = Reactions.toggled(prev, emoji: emoji, me: d.me.id, on: on)
        upsertLocal(next)
        let path = "/messages/\(m.id)/reactions/\(emoji.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? emoji)"
        var body: [String: Any]? = nil
        if on {
            body = [:]
            if emoji == Reactions.look && Reactions.actionsEnabled(d) { body?["remindAt"] = ISODate.string(Reactions.lookRemindAt()) }
        }
        do {
            let r: ReactionResult = try await api.request(path, method: on ? "PUT" : "DELETE", json: body)
            if let msg = r.message { upsertLocal(msg) }
            if r.reminder != nil || !r.closedReminderIds.isEmpty {
                let closed = Set(r.closedReminderIds)
                reminders = (reminders.filter { !closed.contains($0.id) && $0.id != r.reminder?.id } + [r.reminder].compactMap { $0 })
                    .sorted { $0.remindAt < $1.remindAt }
            }
            return r
        } catch {
            var back = conversations[m.conversationId]?.messages.first { $0.id == m.id } ?? current
            back.reactions = prev
            upsertLocal(back)
            throw error
        }
    }
}

File: apps/android/app/src/main/java/com/tiecoms/app/core/Reactions.kt
package com.tiecoms.app.core

import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * Reacciones con emoji (docs/REACCIONES_ENLACES.md; mismas reglas en web, iOS y Android).
 * Sin dependencias de Android: se prueba en la JVM. Las tablas de Unicode son las de emoji-data.txt (Emoji 15.1);
 * el servidor vuelve a normalizar de todos modos, así que un emoji más nuevo que estas tablas igual pasa.
 */
object Reactions {
    /** Barra rápida, en este orden (QUICK_REACTIONS del contrato). */
    val QUICK = listOf("👍", "❤️", "😂", "👀", "✅", "🙏")
    const val LOOK = "👀"
    const val DONE = "✅"
    /** Máximo de emojis distintos por mensaje (MAX_REACTIONS_PER_MESSAGE): pasado eso el API responde 409. */
    const val MAX_PER_MESSAGE = 20

    /** Selector completo: los de uso diario en el trabajo primero; el teclado del sistema trae el resto. */
    val PICKER = listOf(
        "👍", "❤️", "😂", "👀", "✅", "🙏", "👏", "🙌", "🎉", "🔥", "💯", "💪",
        "😀", "😃", "😄", "😁", "😅", "🤣", "😊", "🙂", "😉", "😍", "🥰", "😘",
        "😎", "🤩", "🥳", "😇", "🤗", "🤔", "🤨", "😐", "😑", "🙄", "😏", "😬",
        "😮", "😯", "😲", "😳", "🥺", "😢", "😭", "😤", "😡", "🤯", "😱", "😴",
        "🤝", "👌", "✌️", "🤞", "👋", "✋", "👆", "👇", "👉", "👈", "🫡", "🫶",
        "👎", "🤦", "🤷", "🙋", "💁", "🙆", "🙅", "🧠", "💡", "📌", "📎", "📝",
        "📅", "⏰", "⌛", "🚀", "⭐", "✨", "⚡", "☕", "🍻", "🎂", "🎁", "🏆",
        "💰", "📈", "📉", "📊", "💼", "🏠", "🚗", "✈️", "📞", "💬", "📣", "🔔",
        "❌", "⚠️", "❓", "❗", "➕", "➖", "🆗", "🆕", "🟢", "🟡", "🔴", "⚪",
        "💚", "💙", "💛", "🧡", "💜", "🖤", "🤍", "💔", "🇨🇴", "🇲🇽", "🇵🇪", "🇪🇸",
    )

    private const val VS16 = 0xFE0F
    private const val ZWJ = 0x200D
    private const val KEYCAP = 0x20E3

    // Extended_Pictographic (emoji-data.txt).
    private val PICTO = intArrayOf(
        0x00A9, 0x00A9, 0x00AE, 0x00AE, 0x203C, 0x203C, 0x2049, 0x2049, 0x2122, 0x2122, 0x2139, 0x2139,
        0x2194, 0x2199, 0x21A9, 0x21AA, 0x231A, 0x231B, 0x2328, 0x2328, 0x2388, 0x2388, 0x23CF, 0x23CF,
        0x23E9, 0x23F3, 0x23F8, 0x23FA, 0x24C2, 0x24C2, 0x25AA, 0x25AB, 0x25B6, 0x25B6, 0x25C0, 0x25C0,
        0x25FB, 0x25FE, 0x2600, 0x2605, 0x2607, 0x2612, 0x2614, 0x2685, 0x2690, 0x2705, 0x2708, 0x2712,
        0x2714, 0x2714, 0x2716, 0x2716, 0x271D, 0x271D, 0x2721, 0x2721, 0x2728, 0x2728, 0x2733, 0x2734,
        0x2744, 0x2744, 0x2747, 0x2747, 0x274C, 0x274C, 0x274E, 0x274E, 0x2753, 0x2755, 0x2757, 0x2757,
        0x2763, 0x2767, 0x2795, 0x2797, 0x27A1, 0x27A1, 0x27B0, 0x27B0, 0x27BF, 0x27BF, 0x2934, 0x2935,
        0x2B05, 0x2B07, 0x2B1B, 0x2B1C, 0x2B50, 0x2B50, 0x2B55, 0x2B55, 0x3030, 0x3030, 0x303D, 0x303D,
        0x3297, 0x3297, 0x3299, 0x3299,
        0x1F000, 0x1F0FF, 0x1F10D, 0x1F10F, 0x1F12F, 0x1F12F, 0x1F16C, 0x1F171, 0x1F17E, 0x1F17F,
        0x1F18E, 0x1F18E, 0x1F191, 0x1F19A, 0x1F1AD, 0x1F1E5, 0x1F201, 0x1F20F, 0x1F21A, 0x1F21A,
        0x1F22F, 0x1F22F, 0x1F232, 0x1F23A, 0x1F23C, 0x1F23F, 0x1F249, 0x1F3FA, 0x1F400, 0x1F53D,
        0x1F546, 0x1F64F, 0x1F680, 0x1F6FF, 0x1F774, 0x1F77F, 0x1F7D5, 0x1F7FF, 0x1F80C, 0x1F80F,
        0x1F848, 0x1F84F, 0x1F85A, 0x1F85F, 0x1F888, 0x1F88F, 0x1F8AE, 0x1F8FF, 0x1F90C, 0x1F93A,
        0x1F93C, 0x1F945, 0x1F947, 0x1FAFF, 0x1FC00, 0x1FFFD,
    )

    // Emoji_Presentation: se ven como emoji sin U+FE0F.
    private val PRESENTATION = intArrayOf(
        0x231A, 0x231B, 0x23E9, 0x23EC, 0x23F0, 0x23F0, 0x23F3, 0x23F3, 0x25FD, 0x25FE, 0x2614, 0x2615,
        0x2648, 0x2653, 0x267F, 0x267F, 0x2693, 0x2693, 0x26A1, 0x26A1, 0x26AA, 0x26AB, 0x26BD, 0x26BE,
        0x26C4, 0x26C5, 0x26CE, 0x26CE, 0x26D4, 0x26D4, 0x26EA, 0x26EA, 0x26F2, 0x26F3, 0x26F5, 0x26F5,
        0x26FA, 0x26FA, 0x26FD, 0x26FD, 0x2705, 0x2705, 0x270A, 0x270B, 0x2728, 0x2728, 0x274C, 0x274C,
        0x274E, 0x274E, 0x2753, 0x2755, 0x2757, 0x2757, 0x2795, 0x2797, 0x27B0, 0x27B0, 0x27BF, 0x27BF,
        0x2B1B, 0x2B1C, 0x2B50, 0x2B50, 0x2B55, 0x2B55,
        0x1F004, 0x1F004, 0x1F0CF, 0x1F0CF, 0x1F18E, 0x1F18E, 0x1F191, 0x1F19A, 0x1F1E6, 0x1F1FF,
        0x1F201, 0x1F201, 0x1F21A, 0x1F21A, 0x1F22F, 0x1F22F, 0x1F232, 0x1F236, 0x1F238, 0x1F23A,
        0x1F250, 0x1F251, 0x1F300, 0x1F320, 0x1F32D, 0x1F335, 0x1F337, 0x1F37C, 0x1F37E, 0x1F393,
        0x1F3A0, 0x1F3CA, 0x1F3CF, 0x1F3D3, 0x1F3E0, 0x1F3F0, 0x1F3F4, 0x1F3F4, 0x1F3F8, 0x1F43E,
        0x1F440, 0x1F440, 0x1F442, 0x1F4FC, 0x1F4FF, 0x1F53D, 0x1F54B, 0x1F54E, 0x1F550, 0x1F567,
        0x1F57A, 0x1F57A, 0x1F595, 0x1F596, 0x1F5A4, 0x1F5A4, 0x1F5FB, 0x1F64F, 0x1F680, 0x1F6C5,
        0x1F6CC, 0x1F6CC, 0x1F6D0, 0x1F6D2, 0x1F6D5, 0x1F6D7, 0x1F6DC, 0x1F6DF, 0x1F6EB, 0x1F6EC,
        0x1F6F4, 0x1F6FC, 0x1F7E0, 0x1F7EB, 0x1F7F0, 0x1F7F0, 0x1F90C, 0x1F93A, 0x1F93C, 0x1F945,
        0x1F947, 0x1F9FF, 0x1FA70, 0x1FA7C, 0x1FA80, 0x1FA89, 0x1FA8F, 0x1FAC6, 0x1FACE, 0x1FADC,
        0x1FADF, 0x1FAE9, 0x1FAF0, 0x1FAF8,
    )

    private fun inRanges(t: IntArray, cp: Int): Boolean {
        var i = 0
        while (i < t.size) { if (cp < t[i]) return false; if (cp <= t[i + 1]) return true; i += 2 }
        return false
    }
    fun isPictographic(cp: Int) = inRanges(PICTO, cp)
    private fun isPresentation(cp: Int) = inRanges(PRESENTATION, cp)
    private fun isSkin(cp: Int) = cp in 0x1F3FB..0x1F3FF
    private fun isRegional(cp: Int) = cp in 0x1F1E6..0x1F1FF
    private fun isTag(cp: Int) = cp in 0xE0020..0xE007F
    private fun isKeycapBase(cp: Int) = cp in '0'.code..'9'.code || cp == '#'.code || cp == '*'.code
    /** Pelirrojo, rizos, calvo, canas: van después de un ZWJ. */
    private fun isHairComponent(cp: Int) = cp in 0x1F9B0..0x1F9B3

    private fun cps(s: String): IntArray = s.codePoints().toArray()
    private fun str(cp: IntArray, from: Int = 0, to: Int = cp.size) = String(cp, from, to - from)

    /**
     * Forma canónica (normalizeEmoji del contrato): sin selectores de variación sobrantes y con U+FE0F donde hace
     * falta para verse como emoji (❤ → ❤️, 👍️ → 👍, 1⃣ → 1️⃣). null si no es exactamente un emoji.
     */
    fun normalize(input: String): String? {
        val raw = input.trim()
        if (raw.isEmpty() || raw.length > 32) return null
        val parts = raw.replace("️", "").split("‍").map { part ->
            val p = cps(part)
            if (p.size == 2 && isKeycapBase(p[0]) && p[1] == KEYCAP) return@map str(intArrayOf(p[0], VS16, KEYCAP))
            if (p.isEmpty()) return@map part
            val first = p[0]
            val needs = isPictographic(first) && !isPresentation(first) && !(p.size > 1 && isSkin(p[1]))
            if (needs) str(intArrayOf(first, VS16) + p.copyOfRange(1, p.size)) else part
        }
        val out = parts.joinToString("‍")
        if (clusters(out)?.size != 1) return null
        val o = cps(out)
        if (o.none { isPictographic(it) || isRegional(it) || it == KEYCAP }) return null
        return out
    }

    /**
     * Parte un texto que debería ser solo emojis en sus emojis (grafemas). null si hay algo que no es emoji.
     * Reglas de UAX #29 reducidas a lo que hace falta aquí: banderas (dos indicadores regionales), keycaps,
     * secuencias con ZWJ, tonos de piel y banderas de subdivisión (etiquetas).
     */
    fun clusters(s: String): List<String>? {
        val c = cps(s)
        if (c.isEmpty()) return null
        val out = mutableListOf<String>()
        var i = 0
        fun elementEnd(start: Int): Int {
            var j = start
            val cp = c[j]
            if (isKeycapBase(cp)) {
                j++
                if (j < c.size && c[j] == VS16) j++
                return if (j < c.size && c[j] == KEYCAP) j + 1 else -1
            }
            if (!isPictographic(cp) && !isSkin(cp) && !isHairComponent(cp)) return -1
            j++
            if (j < c.size && c[j] == VS16) j++
            if (j < c.size && isSkin(c[j])) j++
            if (j < c.size && c[j] == VS16) j++
            while (j < c.size && isTag(c[j])) j++
            return j
        }
        while (i < c.size) {
            val start = i
            if (isRegional(c[i])) {
                if (i + 1 >= c.size || !isRegional(c[i + 1])) return null
                i += 2
                out += str(c, start, i); continue
            }
            var j = elementEnd(i)
            if (j < 0) return null
            while (j < c.size && c[j] == ZWJ && j + 1 < c.size) {
                val k = elementEnd(j + 1)
                if (k < 0) return null
                j = k
            }
            i = j
            out += str(c, start, i)
        }
        return out
    }

    /** Solo emojis (de 1 a 3, sin contar espacios): el mensaje se muestra grande (isJumbo de la web). */
    fun isJumbo(body: String): Boolean {
        val s = body.trim()
        if (s.isEmpty() || s.length > 40) return false
        val n = clusters(s.filterNot { it.isWhitespace() })?.size ?: return false
        return n in 1..3
    }

    /** Mi reacción puesta o quitada (la actualización optimista, igual que client.react de la web). */
    fun toggle(list: List<ReactionDTO>, emoji: String, me: String, on: Boolean): List<ReactionDTO> {
        var next = list.map { r -> if (r.emoji == emoji) r.copy(userIds = r.userIds.filter { it != me }) else r }
        if (on) {
            next = if (next.any { it.emoji == emoji }) next.map { r -> if (r.emoji == emoji) r.copy(userIds = r.userIds + me) else r }
            else next + ReactionDTO(emoji, listOf(me))
        }
        return next.filter { it.userIds.isNotEmpty() || it.external.isNotEmpty() }
    }

    fun mine(r: ReactionDTO, me: String) = me in r.userIds

    /** ¿Cabe otro emoji distinto? (el máximo es por emoji nuevo; sumarse a uno que ya está siempre se puede). */
    fun canAdd(list: List<ReactionDTO>, emoji: String) = list.any { it.emoji == emoji } || list.count { it.count > 0 } < MAX_PER_MESSAGE

    /** Recordatorio de 👀: en 3 horas, o mañana a las 9:00 si eso cae de noche (19:00 o después) o en otro día. */
    fun lookRemindAt(now: ZonedDateTime = ZonedDateTime.now(ZoneId.systemDefault())): ZonedDateTime {
        val at = now.plusHours(3)
        if (at.hour >= 19 || at.toLocalDate() != now.toLocalDate()) {
            return ZonedDateTime.of(LocalDateTime.of(now.toLocalDate().plusDays(1), java.time.LocalTime.of(9, 0)), now.zone)
        }
        return at
    }
}
